### PR TITLE
Support primitive arguments of agents in cli scan run

### PR DESCRIPTION
**Pull request by benyissa**
Example of usage:

image

Let's test the nmap agent with the `top_ports` argument, which is an integer:

Case 1: nmap with top_ports arg added correctly

```shell
oxo scan run --agent agent/ostorlab/nmap --follow agent/ostorlab/nmap --arg top_ports:1 ip 8.8.8.8
```
The scan logs show that the agent is correctly running with `top_ports = 1`.

image

Case 2: `top_ports` with wrong value, not an integer:

```shell
oxo scan run --agent agent/ostorlab/nmap --follow agent/ostorlab/nmap --arg top_ports:test ip 8.8.8.8
```

image

Case 3: Unsupported format

image

Case 4 : Agent image not found in local

image

### CODE REVIEW

File: src/ostorlab/cli/agent_fetcher.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+from ostorlab import configuration_manager
+from ostorlab.apis import agent_details as agent_details_api
+from ostorlab.apis.runners import public_runner, authenticated_runner
+from ostorlab.apis.runners import runner as base_runner
+from ostorlab.cli import console as cli_console
+from ostorlab.agent import definitions as agent_definitions
+
+import click
+import docker
+import docker.errors
```
Comment (rabsondev) [inline src/ostorlab/cli/agent_fetcher.py:16]:
```suggestion
import logging
import io
from typing import Any

import click
import docker
import docker.errors

from ostorlab import configuration_manager
from ostorlab.apis import agent_details as agent_details_api
from ostorlab.apis.runners import public_runner, authenticated_runner
from ostorlab.apis.runners import runner as base_runner
from ostorlab.cli import console as cli_console
from ostorlab.agent import definitions as agent_definitions
```

File: src/ostorlab/cli/scan/run/run.py [hunk 2]
```diff
@@ -135,11 +148,39 @@ def run(
...
+                                            value=cli_argument.value,
+                                        )
+                                    )
+                                except ValueError as e:
+                                    console.error(f"{e}")
+                                    raise click.ClickException(
+                                        f"{cli_argument.value} is not a Valid value for {cli_argument.name}. please "
+                                        f"check the agent documentation for the correct value type. at "
+                                        f"https://github.com/ostorlab/agent_{agent_setting.key.split('/')[-1]}/blob"
+                                        f"/main/ostorlab.yaml"
```
Comment (rabsondev) [inline src/ostorlab/cli/scan/run/run.py:178]:
```suggestion
                                        f"{cli_argument.value} is not a Valid value for {cli_argument.name}. Please "
                                        f"check the agent documentation for the correct value type at "
                                        f"https://github.com/ostorlab/agent_{agent_setting.key.split('/')[-1]}/blob"
                                        f"/main/ostorlab.yaml"
```

File: src/ostorlab/cli/scan/run/run.py [hunk 3]
```diff
@@ -135,11 +148,39 @@ def run(
...
+                                    ) from e
+                except agent_fetcher.AgentDetailsNotFound:
+                    console.warning(
+                        f"Agent {agent_setting.key} not found. use oxo agent install {agent_setting.key} to install it"
```
Comment (rabsondev) [inline src/ostorlab/cli/scan/run/run.py:182]:
```suggestion
                    console.warning(
                        f"Agent {agent_setting.key} not found. Use oxo agent install {agent_setting.key} to install it."
```

File: tests/cli/scan/run/run_test.py [hunk 4]
```diff
@@ -361,3 +361,25 @@ def testScanRunLocalRuntime_whenIInvalidYamlAgentGroupDefinition_showsErrorMessa
     assert isinstance(result.exception, SystemExit)
     assert result.exit_code == 2
     assert "Agent group definition YAML parse error" in result.output
+
+
+def testOstorlabScanRunCLI_whenWrongArgsFormatProvided_showsErrorMessage() -> None:
```
Comment (rabsondev) [inline tests/cli/scan/run/run_test.py:366]: What about the test for the correct args format?

Comment (rabsondev) [inline tests/cli/scan/run/run_test.py:366]: What about the test for the correct args format?

File: src/ostorlab/agent/agent.py [hunk 5]
```diff
@@ -157,7 +157,8 @@ def args(self) -> Dict[str, Any]:
             if a.type == "binary":
                 arguments[a.name] = a.value
             else:
-                arguments[a.name] = json.loads(a.value.decode())
+                arg_value = a.value.decode() if isinstance(a.value, bytes) else a.value
```
Comment (elyousfi5) [inline src/ostorlab/agent/agent.py:160]: is there a possiblity for this to raise UnicodeDecodeError ? if yes we should handle it probably !!

File: src/ostorlab/cli/scan/run/run.py [hunk 6]
```diff
@@ -66,6 +78,7 @@
 def run(
     ctx: click.core.Context,
     agent: List[str],
+    arg: list[types.AgentArg],
```
Comment (elyousfi5) [inline src/ostorlab/cli/scan/run/run.py:81]:
```suggestion
    args: list[types.AgentArg],
```

File: src/ostorlab/cli/scan/run/run.py (hunk 3 above)
Comment (elyousfi5) [inline src/ostorlab/cli/scan/run/run.py:182]:
```suggestion
                        f"Agent {agent_setting.key} not found. Use oxo agent install {agent_setting.key} to install it"
```

File: src/ostorlab/agent/agent.py (hunk 5 above)
Comment (benyissa) [reply src/ostorlab/agent/agent.py:160]: The decode step already exists. so I don't think so. The change here is just to read the value if it wasn't received in bytes.

File: src/ostorlab/cli/agent_fetcher.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+    Returns:
+        dictionary of the agent information like : name, dockerLocation..
+
+    Raises:
+        click Exit exception with satus code 2 when API response is invalid.
+    """
+    config_manager = configuration_manager.ConfigurationManager()
+
+    if config_manager.is_authenticated:
```
[the remark this answers is not in the document]
Comment (oussamaessaji) [reply src/ostorlab/cli/agent_fetcher.py:42]: +1

File: src/ostorlab/cli/agent_fetcher.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    try:
+        response = runner.execute(agent_details_api.AgentDetailsAPIRequest(agent_key))
+    except base_runner.ResponseError as e:
+        raise AgentDetailsNotFound("requested agent not found") from e
+
+    if "errors" in response:
+        error_message = f"""The provided agent key : {agent_key} does not correspond to any agent.
+        Please make sure you have the correct agent key.
+        """
+        raise AgentDetailsNotFound(error_message)
```
Comment (oussamaessaji) [inline src/ostorlab/cli/agent_fetcher.py:56]: I think this check already exists in the runner!

File: src/ostorlab/cli/agent_fetcher.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+) -> agent_definitions.AgentDefinition:
+    """Fetch args of an agent from container image.
+    Args:
+        agent_key: key of the agent in agent/org/agentName format.
+    Returns:
+        agent_definition: AgentDefinition object containing the agent definition.
+    """
+
+    docker_client = docker.from_env()
+    agent_details = get_agent_details(agent_key)
```
Comment (mohsinenar) [inline src/ostorlab/cli/agent_fetcher.py:73]: What will happen if the agent dose does not exist in the ostorlab store?

File: src/ostorlab/cli/scan/run/run.py [hunk 10]
```diff
@@ -135,11 +148,39 @@ def run(
                 for ag in agent_group.agents:
                     try:
                         install_agent.install(ag.key, ag.version)
-                    except install_agent.AgentDetailsNotFound:
+                    except agent_fetcher.AgentDetailsNotFound:
                         console.warning(f"agent {ag.key} not found on the store")
             except httpx.HTTPError as e:
                 raise click.ClickException(f"Could not install the agents: {e}")
-
+        if arg is not None and len(arg) > 0:
```
Comment (mohsinenar) [inline src/ostorlab/cli/scan/run/run.py:155]: This whole section needs its own method!

**Review by mohsinenar — CHANGES_REQUESTED** [review]

File: src/ostorlab/cli/scan/run/run.py [hunk 11]
```diff
@@ -135,11 +148,39 @@ def run(
...
+                                    console.error(f"{e}")
+                                    raise click.ClickException(
+                                        f"{cli_argument.value} is not a Valid value for {cli_argument.name}. Please "
+                                        f"check the agent documentation for the correct value type at "
+                                        f"https://github.com/ostorlab/agent_{agent_setting.key.split('/')[-1]}/blob"
+                                        f"/main/ostorlab.yaml"
+                                    ) from e
+                except agent_fetcher.AgentDetailsNotFound:
+                    console.warning(
+                        f"Agent {agent_setting.key} not found. Use oxo agent install {agent_setting.key} to install it."
```
[the remark this answers is not in the document]
Comment (3asm) [reply src/ostorlab/cli/scan/run/run.py:182]: Don't we already have logic for non found agents.

**Review by 3asm — DISMISSED** [review]

File: src/ostorlab/cli/agent_fetcher.py [hunk 12]
```diff
@@ -0,0 +1,88 @@
+"""Module responsible for fetching the agent details from the container image."""
```
[the remark this answers is not in the document]
Comment (benyissa) [reply src/ostorlab/cli/agent_fetcher.py:1]: I think with the current structure of cli folder it's fine to put it here too.

image

File: src/ostorlab/cli/types.py [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+
+
+@dataclasses.dataclass
+class AgentArg:
+    name: str
+    value: str
+
+
+class AgentArgType(click.ParamType):
+    name = "agent_arg"
```
[the remark this answers is not in the document]
Comment (benyissa) [reply src/ostorlab/cli/types.py:16]: Alright . It's not clear. I changed it to the following 

image

File: src/ostorlab/cli/scan/run/run.py [hunk 14]
```diff
@@ -150,3 +166,53 @@ def run(
...
+
+        for cli_argument in cli_args:
+            supported_arg = next(
+                (
+                    arg
+                    for arg in agent_definition.args
+                    if arg.get("name") == cli_argument.name
+                ),
+                None,
+            )
```
Comment (3asm) [inline src/ostorlab/cli/scan/run/run.py:202]: I really don't like this pattern. It is so unreadable.

File: src/ostorlab/cli/scan/run/run.py [hunk 15]
```diff
@@ -150,3 +166,53 @@ def run(
...
+            if supported_arg is None:
+                continue
+            try:
+                agent_setting.args.append(
+                    utils_definitions.Arg.from_values(
+                        name=cli_argument.name,
+                        type=supported_arg.get("type", "string"),
+                        value=cli_argument.value,
+                    )
+                )
```
Comment (3asm) [inline src/ostorlab/cli/scan/run/run.py:212]: Could have been don in the for loop for the arg instead of doing, checking None and then this code.

**Review by 3asm — DISMISSED** [review]

File: src/ostorlab/runtimes/definitions.py [hunk 16]
```diff
@@ -119,7 +89,7 @@ def from_proto(cls, proto: bytes) -> "AgentSettings":
             bus_management_url=instance.bus_management_url,
             bus_vhost=instance.bus_vhost,
             args=[
-                defintions.Arg(name=a.name, type=a.type, value=a.value)
+                defintions.Arg.from_values(name=a.name, type=a.type, value=a.value)
```
[the remark this answers is not in the document]
Comment (benyissa) [reply src/ostorlab/runtimes/definitions.py]: Values here mean the values of attributes. but it's not clear. I will use build instead.

File: src/ostorlab/cli/agent_fetcher.py [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+
+console = cli_console.Console()
+logger = logging.getLogger(__name__)
+
+
+class Error(Exception):
+    """Base Error."""
+
+
+class AgentDetailsNotFound(Error):
```
Comment (BlueSquare1) [inline src/ostorlab/cli/agent_fetcher.py:27]:
```suggestion
class AgentNotFound(Error):
```

File: src/ostorlab/cli/agent_fetcher.py [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+    """Agent not found error."""
+
+
+def get_details(agent_key: str) -> dict[Any, Any]:
```
Comment (elyousfi5) [inline src/ostorlab/cli/agent_fetcher.py:31]:
```suggestion
def get_details(agent_key: str) -> dict[str, Any]:
```

**Review by 3asm — APPROVED** [review]

**Comment by 3asm** [comment]
Ping on PR.

**Review by 3asm — APPROVED** [review]

**Review by 3asm — APPROVED** [review]

**Pull request merged**